src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs

```diff
@@ -74,18 +74,114 @@ private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator
 
         private T GetUpdatedSymbol<T>(BoundNode expr, T sym) where T : Symbol?
         {
-            if (sym is null) return sym;
+            if (sym is null || _snapshotManager is null) return sym;
 
-            if (_updatedSymbols.TryGetValue((expr, sym), out var updatedSymbol))
+            if (!_snapshotManager.TryGetUpdatedSymbol(expr, sym, out var updatedSymbol))
             {
-                Debug.Assert(updatedSymbol is object);
-                return (T)updatedSymbol;
+                updatedSymbol = sym;
+            }
+            Debug.Assert(updatedSymbol is object);
+
+            switch (updatedSymbol)
+            {
+                case LambdaSymbol lambda:
+                    return (T)remapLambda((BoundLambda)expr, lambda);
+
+                case LocalFunctionSymbol localFunction:
+                    return (T)remapLocalFunction(localFunction);
+
+                case LocalSymbol local:
+                    return (T)remapLocal(local);
+
+                case ParameterSymbol param:
+                    if (_remappedSymbols.TryGetValue(param, out var updatedParam))
+                    {
+                        return (T)updatedParam;
+                    }
+                    break;
             }
 
-            return sym;
+            return (T)updatedSymbol;
+
+            Symbol remapLambda(BoundLambda boundLambda, LambdaSymbol lambda)
+            {
+                var updatedDelegateType = _snapshotManager!.GetUpdatedDelegateTypeForLambda(lambda);
+
+                if (!_remappedSymbols.TryGetValue(lambda.ContainingSymbol, out Symbol? updatedContaining) && updatedDelegateType is null)
+                {
+                    return lambda;
+                }
+
+                LambdaSymbol updatedLambda;
+                if (updatedDelegateType is null)
+                {
+                    Debug.Assert(updatedContaining is object);
+                    var parameterRefKinds = lambda.ParameterRefKinds.IsDefault ? ImmutableArray<RefKind>.Empty : lambda.ParameterRefKinds;
+                    updatedLambda = boundLambda.CreateLambdaSymbol(updatedContaining, lambda.ReturnTypeWithAnnotations, lambda.ParameterTypesWithAnnotations, parameterRefKinds, lambda.RefKind);
+                }
+                else
+                {
+                    Debug.Assert(updatedDelegateType is object);
+                    updatedLambda = boundLambda.CreateLambdaSymbol(updatedDelegateType, updatedContaining ?? lambda.ContainingSymbol);
+                }
+
+                _remappedSymbols.Add(lambda, updatedLambda);
```

Comment:
> _remappedSymbols.Add(lambda, updatedLambda); [](start = 15, length = 45)

Is there a guarantee that we will not try to do this more than once?

Reply:
`Add` will throw if we add attempt to add the same key twice.

---
In reply to: [329752539](https://github.com/dotnet/roslyn/pull/38769#discussion_r329752539) [](ancestors = 329752539)

Reply:
> Add will throw if we add attempt to add the same key twice.

Is there a guarantee that this will not happen?

---
In reply to: [329827877](https://github.com/dotnet/roslyn/pull/38769#discussion_r329827877) [](ancestors = 329827877,329752539)

Reply:
Nothing beyond the standard invariants around how BoundLambdas generate their symbols when asked for them, and that such symbols are unique.

---
In reply to: [330680758](https://github.com/dotnet/roslyn/pull/38769#discussion_r330680758) [](ancestors = 330680758,329827877,329752539)